source/server/overload_manager_impl.cc

```diff
@@ -17,22 +19,91 @@ namespace Server {
 
 namespace {
 
-class ThresholdTriggerImpl : public OverloadAction::Trigger {
+class ThresholdTriggerImpl final : public OverloadAction::Trigger {
 public:
   ThresholdTriggerImpl(const envoy::config::overload::v3::ThresholdTrigger& config)
       : threshold_(config.value()) {}
 
   bool updateValue(double value) override {
-    const bool fired = isFired();
-    value_ = value;
-    return fired != isFired();
+    const OverloadActionState state = actionState();
+    state_ =
+        value >= threshold_ ? OverloadActionState::saturated() : OverloadActionState::inactive();
+    return state != actionState();
   }
 
-  bool isFired() const override { return value_.has_value() && value_ >= threshold_; }
+  OverloadActionState actionState() const override {
+    return state_.value_or(OverloadActionState::inactive());
+  }
 
 private:
   const double threshold_;
-  absl::optional<double> value_;
+  absl::optional<OverloadActionState> state_;
+};
+
+class RangeTriggerImpl final : public OverloadAction::Trigger {
+public:
+  RangeTriggerImpl(const envoy::config::overload::v3::RangeTrigger& config)
+      : minimum_(config.min_value()),
+        maximum_(config.has_max_value() ? config.max_value().value() : 1.0) {
+    if (minimum_ >= maximum_) {
+      throw EnvoyException("min_value must be less than max_value");
+    }
+  }
+
+  bool updateValue(const double value) override {
+    const OverloadActionState state = actionState();
+    if (value <= minimum_) {
+      state_ = OverloadActionState::inactive();
+    } else if (value >= maximum_) {
+      state_ = OverloadActionState::saturated();
+    } else {
+      state_ = OverloadActionState((value - minimum_) / (maximum_ - minimum_));
+    }
+    return state_ != state;
+  }
+
+  OverloadActionState actionState() const override {
+    return state_.value_or(OverloadActionState::inactive());
+  }
+
+private:
+  const double minimum_;
+  const double maximum_;
+  absl::optional<OverloadActionState> state_;
+};
+
+/**
+ * Thread-local copy of the state of each configured overload action.
+ */
+class ThreadLocalOverloadStateImpl : public ThreadLocalOverloadState,
+                                     public ThreadLocal::ThreadLocalObject {
+public:
+  explicit ThreadLocalOverloadStateImpl(Event::Dispatcher& dispatcher) : dispatcher_(dispatcher) {}
+
+  const OverloadActionState& getState(const std::string& action) override {
+    auto it = actions_.find(action);
+    if (it == actions_.end()) {
+      it = actions_.insert(std::make_pair(action, OverloadActionState::inactive())).first;
+    }
+    return it->second;
+  }
+
+  void setState(const std::string& action, OverloadActionState state) override {
+    auto insert = actions_.emplace(action, state);
+    if (!insert.second) {
+      insert.first->second = state;
+    }
+  }
+
+  OverloadTimerFactory getTimerFactory() override {
+    return [this](absl::string_view /*unused*/, std::function<void()> callback) {
```

Comment:
Use of an enum type that we can cast to to int would provide more efficient lookup than string_view.

Reply:
Do we want this to be available as an extension point? I could see implementers of custom filters and such wanting to register their own timers for scaling in response to load.

Reply:
Does it matter what kind of timer is being created?  The actual timeout value is an input to the enable method.

What if Timer had a method enableTimerRange(const std::chrono::milliseconds& min, const std::chrono::milliseconds& max) that scales timeout based on overload manager state?